crates/levm/tests/tests.rs

```diff
@@ -25,3 +25,174 @@ fn test() {
+    stored_value.to_big_endian(&mut value_bytes);
+
+    assert_eq!(value_bytes[0..1], [0xAB]);
+}
+
+#[test]
+fn mcopy() {
+    let mut vm = VM::default();
+
+    vm.stack.push(U256::from(32)); // size
+    vm.stack.push(U256::from(0)); // source offset
+    vm.stack.push(U256::from(64)); // destination offset
+
+    vm.stack.push(U256::from(0x33333)); // value
+    vm.stack.push(U256::from(0)); // offset
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MCOPY as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let copied_value = vm.memory.load(64);
+    assert_eq!(copied_value, U256::from(0x33333));
+}
+
+#[test]
+fn mload() {
+    let mut vm = VM::default();
+
+    vm.stack.push(U256::from(0)); // offset to load
+
+    vm.stack.push(U256::from(0x33333)); // value to store
+    vm.stack.push(U256::from(0)); // offset to store
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MLOAD as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let loaded_value = vm.stack.pop().unwrap();
+    assert_eq!(loaded_value, U256::from(0x33333));
+}
+
+#[test]
+fn msize() {
+    let mut vm = VM::default();
+
+    vm.execute(Bytes::from(vec![Opcode::MSIZE as u8, Opcode::STOP as u8]));
+    let initial_size = vm.stack.pop().unwrap();
+    assert_eq!(initial_size, U256::from(0));
+
+    vm.pc = 0;
+
+    vm.stack.push(U256::from(0x33333)); // value
+    vm.stack.push(U256::from(0)); // offset
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MSIZE as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let after_store_size = vm.stack.pop().unwrap();
+    assert_eq!(after_store_size, U256::from(32));
+
+    vm.pc = 0;
+
+    vm.stack.push(U256::from(0x55555)); // value
+    vm.stack.push(U256::from(64)); // offset
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MSIZE as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let final_size = vm.stack.pop().unwrap();
+    assert_eq!(final_size, U256::from(96));
+}
+
+#[test]
+fn mstore_mload_offset_not_multiple_of_32() {
+    let mut vm = VM::default();
+
+    vm.stack.push(U256::from(10)); // offset
+
+    vm.stack.push(U256::from(0xabcdef)); // value
+    vm.stack.push(U256::from(10)); // offset
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MLOAD as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let loaded_value = vm.stack.pop().unwrap();
+    assert_eq!(loaded_value, U256::from(0xabcdef));
+
+    //check with big offset
+
+    vm.pc = 0;
+
+    vm.stack.push(U256::from(2000)); // offset
+
+    vm.stack.push(U256::from(0x123456)); // value
+    vm.stack.push(U256::from(2000)); // offset
+
+    vm.execute(Bytes::from(vec![
+        Opcode::MSTORE as u8,
+        Opcode::MLOAD as u8,
+        Opcode::STOP as u8,
+    ]));
+
+    let loaded_value = vm.stack.pop().unwrap();
+    assert_eq!(loaded_value, U256::from(0x123456));
```

Comment:
Let's check memory was expanded correctly. In the first case it should be 64, and in the second one it should be 2048

Reply:
done!